inbm/dispatcher-agent/dispatcher/source/ubuntu_source_manager_genai_unit_test.py

```diff
@@ -0,0 +1,363 @@
+sys.path.insert(0, '/home/runner/GITHUB_ACTION_RUNNERS/_work/genai_ag_intel-inb-manageability/genai_ag_intel-inb-manageability/inbm-lib/')
+sys.path.insert(0, '/home/runner/GITHUB_ACTION_RUNNERS/_work/genai_ag_intel-inb-manageability/genai_ag_intel-inb-manageability/inbm/dispatcher-agent/dispatcher/source/')
+import pytest
+from unittest.mock import patch, MagicMock
+from dispatcher.source.ubuntu_source_manager import UbuntuApplicationSourceManager, ApplicationAddSourceParameters
+from unittest.mock import patch, mock_open
+from dispatcher.source.ubuntu_source_manager import UbuntuApplicationSourceManager, ApplicationSourceList
+from dispatcher.source.ubuntu_source_manager import UbuntuApplicationSourceManager, ApplicationRemoveSourceParameters, SourceError
+from dispatcher.source.ubuntu_source_manager import UbuntuApplicationSourceManager, SourceError, ApplicationUpdateSourceParameters
+from dispatcher.source.ubuntu_source_manager import UbuntuOsSourceManager, SourceParameters, SourceError
+from dispatcher.source.ubuntu_source_manager import UbuntuOsSourceManager, SourceError
+
+#DO NOT DELETE THIS LINE - TestUbuntuApplicationSourceManagerAdd
+'''
+ADD HUMAN FEEDBACK BELOW:
+
+'''
+class TestUbuntuApplicationSourceManagerAdd:
+    @pytest.fixture
+    def mock_dispatcher_broker(self):
+        return MagicMock()
+
+    @pytest.fixture
+    def mock_application_source_manager(self, mock_dispatcher_broker):
+        return UbuntuApplicationSourceManager(mock_dispatcher_broker)
+
+    @pytest.mark.parametrize('gpg_key_name, gpg_key_uri, source_list_file_name, sources', [
+        ('test_key', 'http://test.com/key', 'test_file', ['deb http://test.com/ubuntu bionic main']),
+        (None, None, 'test_file', ['deb http://test.com/ubuntu bionic main']),
+        ('test_key', 'http://test.com/key', 'test_file', []),
+    ])
+    @patch('dispatcher.source.ubuntu_source_manager.verify_source')
+    @patch('dispatcher.source.ubuntu_source_manager.add_gpg_key')
+    @patch('dispatcher.source.ubuntu_source_manager.create_file_with_contents')
+    def test_add(self, mock_create_file_with_contents, mock_add_gpg_key, mock_verify_source, mock_application_source_manager, gpg_key_name, gpg_key_uri, source_list_file_name, sources):
+        # Arrange
+        parameters = ApplicationAddSourceParameters(gpg_key_name=gpg_key_name, gpg_key_uri=gpg_key_uri, source_list_file_name=source_list_file_name, sources=sources)
+
+        # Act
+        mock_application_source_manager.add(parameters)
+
+        # Assert
+        if gpg_key_name and gpg_key_uri:
+            mock_verify_source.assert_called_once()
+            mock_add_gpg_key.assert_called_once_with(gpg_key_uri, gpg_key_name)
+        mock_create_file_with_contents.assert_called_once_with('/etc/apt/sources.list.d/' + source_list_file_name, sources)
+
+#DO NOT DELETE THIS LINE - TestUbuntuApplicationSourceManagerList
+'''
+ADD HUMAN FEEDBACK BELOW:
+
+'''
+class TestUbuntuApplicationSourceManagerList:
```

Comment:
To go back to Pull Request Summary table, Click on the [link](https://github.com/ankit-r-gajera/genai_ag_intel-inb-manageability/pull/4#)

<h3>Class: UbuntuApplicationSourceManager |Function: list</h3>

[Source Code Link](https://github.com/ankit-r-gajera/genai_ag_intel-inb-manageability/blob/NEX_MAXIMUS_UT/develop/inbm/dispatcher-agent/dispatcher/source/ubuntu_source_manager.py?plain=1#L125)

[Unit Test Case Code Link](https://github.com/ankit-r-gajera/genai_ag_intel-inb-manageability/blob/NEX_MAXIMUS_UT/develop/inbm/dispatcher-agent/dispatcher/source/ubuntu_source_manager_genai_unit_test.py?plain=1#L58)

<h4>Test Case Scenarios:</h4>
The unit test `test_list` in the `TestUbuntuApplicationSourceManagerList` class generates four test cases using the `pytest.mark.parametrize` decorator. Here are the test cases along with their explanations:

1. Test Case: `test_list(["deb http://archive.ubuntu.com/ubuntu/ focal universe"], [ApplicationSourceList(name="sources.list", sources=["deb http://archive.ubuntu.com/ubuntu/ focal universe"])])`
   - Explanation: This test case tests the scenario where the source file contains one source. The expected output is a list with one `ApplicationSourceList` object that contains the source.

2. Test Case: `test_list(["deb http://archive.ubuntu.com/ubuntu/ focal universe", "deb-src http://archive.ubuntu.com/ubuntu/ focal universe"], [ApplicationSourceList(name="sources.list", sources=["deb http://archive.ubuntu.com/ubuntu/ focal universe", "deb-src http://archive.ubuntu.com/ubuntu/ focal universe"])])`
   - Explanation: This test case tests the scenario where the source file contains multiple sources. The expected output is a list with one `ApplicationSourceList` object that contains all the sources.

3. Test Case: `test_list([], [ApplicationSourceList(name="sources.list", sources=[])])`
   - Explanation: This test case tests the scenario where the source file is empty. The expected output is a list with one `ApplicationSourceList` object with an empty `sources` list.

4. Test Case: `test_list(["# This is a comment", "", "deb http://archive.ubuntu.com/ubuntu/ focal universe"], [ApplicationSourceList(name="sources.list", sources=["deb http://archive.ubuntu.com/ubuntu/ focal universe"])])`
   - Explanation: This test case tests the scenario where the source file contains comments and empty lines in addition to sources. The expected output is a list with one `ApplicationSourceList` object that contains only the sources, ignoring the comments and empty lines.

Each test case calls the `list` method of the `UbuntuApplicationSourceManager` class and asserts that the output matches the expected output. The `glob.glob` and `builtins.open` functions are mocked to control their behavior during the test.


<h4>Test Execution Results:</h4>

| Total Cases    | 4  |
| -------------- | -- |
| Passed Cases   | 4  |
| Failed Cases   | 0  |



To go back to Pull Request Summary table, Click on the [link](https://github.com/ankit-r-gajera/genai_ag_intel-inb-manageability/pull/4#)